### PR TITLE
allow escaped null in urls

### LINKED ISSUE #471: test nulls and escaped nulls in encoded strings
Like the params and segments, but also everything else

### PR DESCRIPTION
**Pull request by alandefreitas**
fix #471

### CODE REVIEW

File: include/boost/url/rfc/impl/query_rule.ipp [hunk 1]
```diff
@@ -64,13 +64,6 @@ parse(
                 BOOST_URL_RETURN_EC(
                     error::bad_pct_hexdig);
             }
-            if (it[1] == '0' &&
-                it[2] == '0')
-            {
-                // null in input
-                BOOST_URL_RETURN_EC(
-                    error::illegal_null);
```
Comment (vinniefalco) [inline include/boost/url/rfc/impl/query_rule.ipp:72]: Are nulls allowed in `application/x-www-form-urlencoded`?

File: test/unit/segments_view.cpp [hunk 2]
```diff
@@ -131,10 +131,66 @@ struct segments_view_test
         }
     }
 
+    void
+    testNull()
```
Comment (vinniefalco) [inline test/unit/segments_view.cpp:135]: This is the wrong file you are spreading stuff having to do with queries, paths, fragments, etc.. all in this file which has to do with segments.

File: include/boost/url/rfc/impl/query_rule.ipp (hunk 1 above)
Comment (alandefreitas) [reply include/boost/url/rfc/impl/query_rule.ipp:72]: I don't know. I would guess it's valid in certain cases. But the problem, in this case, is that query_rule is not being used for `application/x-www-form-urlencoded` only. It's used for queries in general and `url`s with nulls cannot be parsed unless `query_rule` accepts nulls. Organizing the query into key/values is also common in schemes other than http.

Comment (vinniefalco) [reply include/boost/url/rfc/impl/query_rule.ipp:72]: > I don't know.

Well what does the spec say

Comment (alandefreitas) [reply include/boost/url/rfc/impl/query_rule.ipp:72]: The URI spec says it's OK. The HTTP spec doesn't say anything.

File: test/unit/url.cpp [hunk 3]
```diff
@@ -1077,6 +1077,61 @@ struct url_test
...
+        BOOST_TEST_EQ(h, eh);
+        BOOST_TEST(stl_equal(h, eh));
+
+        std::string p = u.path();
+        BOOST_TEST_NE(p, "/pa");
+        BOOST_TEST_NE(p, "/pa\0th");
+        std::string ep("/pa\0th");
+        BOOST_TEST_NE(p, ep);
+        BOOST_TEST_NOT(stl_equal(p, ep));
+        ep.assign({'/', 'p', 'a', '\0', 't', 'h'});
```
Comment (vinniefalco) [inline test/unit/url.cpp:1110]: Could you instead write
```
ep = std::string( "/pa\0th", 6 );
```
?

**Comment by vinniefalco** [comment]
rebase and merge

**Pull request merged**